src/main/java/io/supertokens/webserver/api/session/SessionAPI.java

```diff
@@ -70,14 +70,24 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
         assert enableAntiCsrf != null;
         JsonObject userDataInJWT = InputParser.parseJsonObjectOrThrowError(input, "userDataInJWT", false);
         assert userDataInJWT != null;
+
         JsonObject userDataInDatabase = InputParser.parseJsonObjectOrThrowError(input, "userDataInDatabase", false);
         assert userDataInDatabase != null;
 
+        JsonObject grantPayload = super.getVersionFromRequest(req).equals("2.13")
+                ? InputParser.parseJsonObjectOrThrowError(input, "grants", false)
+                : null;
+
         try {
-            SessionInformationHolder sessionInfo = Session.createNewSession(main, userId, userDataInJWT,
+            SessionInformationHolder sessionInfo = Session.createNewSession(main, userId, userDataInJWT, grantPayload,
                     userDataInDatabase, enableAntiCsrf);
 
             JsonObject result = new JsonParser().parse(new Gson().toJson(sessionInfo)).getAsJsonObject();
+            if (!super.getVersionFromRequest(req).equals("2.13")) {
+                result.getAsJsonObject("session").remove("grants");
+            } else if (!result.getAsJsonObject("session").has("grants")) {
+                result.getAsJsonObject("session").add("grants", new JsonObject());
```

Comment:
Ideally we should have normalised session grant when fetching from db etc..

Reply:
This grants object only appears on the API level, though, and it's only necessary when getting a V2 session with the new CDI version.